Sprint 8 solution time and duration

churmantey opened the pull request:
На ревью.

churmantey committed:
Добавлены поля времени начала и продолжительности
Расчет времени и продолжительности эпиков

churmantey committed:
Добавлены списки задач упорядоченные по приоритету (дате начала)

Добавлены варианты конструкторов с указанием времени начала и продолжительности

Учтено время и продолжительность при сохранении/чтении из файла

churmantey committed:
Приоритизация задач по времени
Контроль пересечений по времени
Покрытие тестами

churmantey committed:
Приоритизация задач по времени
Контроль пересечений по времени
Покрытие тестами

TyutterinYakov commented on src/dc/yandex/kanban/model/Task.java:
```diff
@@ -8,13 +12,30 @@ public class Task {
     private String description;
     private TaskStatus status;
     protected TaskType type;
```
Можно было бы поступить проще) Вместо того, чтобы вводить поле, можно было бы в классе `Task` объявить метод `public TaskType getType() { return TaskType.TASK; }`
А в классах наследниках Epic, Subtask мы можем переопределить
`@Override
public TaskType getType() { return TaskType.SUBTASK; }`
Аналогично для эпика)
Таким образом, в зависимости от того, объект какого класса был создан, будет возвращаться конкретный тип при вызове этого метода)

churmantey replied:
Логично! Исправлю.

TyutterinYakov commented on src/dc/yandex/kanban/model/Task.java:
```diff
@@ -54,6 +75,26 @@ public void setStatus(TaskStatus status) {
+    }
+
+    public LocalDateTime getEndTime() {
+        return startTime.plus(duration);
```
Не забываем также проверить, что оба значения не `null`, в противном случае следует вернуть `null`, чтобы не получить NPE)

churmantey replied:
Точно.

TyutterinYakov commented on src/dc/yandex/kanban/model/TaskStartTimeComparator.java:
```diff
@@ -0,0 +1,12 @@
+
+public class TaskStartTimeComparator implements Comparator<Task> {
+    @Override
+    public int compare(Task o1, Task o2) {
```
Если время старта не задано - такие задачи должны находиться в конце списка, а их сортировка будет осуществлена по возрастанию идентификатора

churmantey replied:
Исправил.

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -41,24 +52,19 @@ public ArrayList<SubTask> getSubTasks() {
+    // и рассчитывает продолжительность
+    public void updateStatusAndTime() {
+        // флаг наличия подзадач в статусе NEW
+        boolean hasNew = subTasks.values().stream()
```
Хорошо было бы реализовать логику вычисления статуса в одном цикле, чтобы не делать это в трех отдельных, что не совсем оптимально

churmantey replied:
Исправил

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -67,13 +73,49 @@ public void updateStatus() {
     // Отключаем самостоятельное обновление статуса эпика пустым методом
     @Override
     public void setStatus(TaskStatus status) {
     }
```
Можно было бы здесь пробросить исключение, чтобы явно указать, что ручное обновление недопустимо)

churmantey replied:
Сделал.

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -67,13 +73,49 @@ public void updateStatus() {
+    // Отключаем самостоятельное обновление времени начала эпика пустым методом
+    @Override
+    public void setStartTime(LocalDateTime startTime) {
+    }
```
Можно было бы здесь пробросить исключение, чтобы явно указать, что ручное обновление недопустимо)

churmantey replied:
Сделал )

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -67,13 +73,49 @@ public void updateStatus() {
+
+    // Отключаем самостоятельное обновление продолжительности
+    @Override
+    public void setDuration(Duration duration) {
```
Можно было бы здесь пробросить исключение, чтобы явно указать, что ручное обновление недопустимо)

churmantey replied:
Сделал.

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -67,13 +73,49 @@ public void updateStatus() {
+        endTime = emptyDate;
+        duration = Duration.ZERO;
+
+        subTasks.values().stream()
```
Аналогично и с временем старта, окончания и продолжительностью) Лучше реализовать в одном цикле

Также хорошо было бы не забыть проверить, что время старта/окончания/продолжительность не `null`

churmantey replied:
Исправил.

TyutterinYakov commented on src/dc/yandex/kanban/model/Epic.java:
```diff
@@ -1,38 +1,49 @@
 
 public class Epic extends Task {
 
     private final HashMap<Integer, SubTask> subTasks; // Список подзадач эпика
```
Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс Map?) Все места не отмечал, но думаю, что ты разберешься)

churmantey replied:
Исправил.)

TyutterinYakov commented on src/dc/yandex/kanban/model/Task.java:
```diff
@@ -8,13 +12,30 @@ public class Task {
+    protected LocalDateTime startTime;
+    protected Duration duration;
+
+    public static final LocalDateTime emptyDate = LocalDateTime.ofInstant(
```
Для большей явности лучше работать с `null`. Есть есть, если дата старта отсутствует - представить ее в виде `null`

churmantey replied:
Принято. Исправил.

TyutterinYakov commented on src/dc/yandex/kanban/service/InMemoryTaskManager.java:
```diff
@@ -3,34 +3,59 @@
+        LocalDateTime t1End = t1.getEndTime();
+        LocalDateTime t2Start = t2.getStartTime();
+        LocalDateTime t2End = t2.getEndTime();
+        return (!t1Start.isBefore(t2Start) && !t1Start.isAfter(t2End))
```
(предложение) Можно поступить проще и отталкиваться не от тех ситуаций, когда все плохо и стоит выбрасывать исключение, а наоборот) Например, время окончания новой до либо равно времени старта существующей - это `(!endTime.isAfter(existStartTime))`. Так же следует добавить подобное условие для второго исхода. В обоих if при true нужно будет просто сделать `continue;` (пропустить итерацию). А, если оба условия будут false, после них можно выбросить исключение)
Так же следует в цикл добавить такое условие)
```
            if (task.getId() == taskPriority.getId()) {
                continue;
            }
```
Так как, если придет задача/подзадача из метода обновления, не факт, что ее время начала/продолжительность будут отличаться от тех, что были, а значит возникнет ошибка валидации) Чтобы это предотвратить, следует пропустить эту итерацию, так как в любом случае старая версия будет удалена из отсортированного списка, если валидация с остальным задачи будет пройдена)
Также следует добавить такую проверку) Так как, если значение времени старта у переданной на валидацию null - это уже означает, что все окей и валидация не требуется. Либо же, если у уде добавленной время старта == null, значит мы дошли до конца, где все остальные будут также иметь время старта null и продолжать проверку нет смысла)
```
if (task.getStartTime() == null || o.getStartTime() == null) {
    return;
}
```

churmantey replied:
Да, согласен. Условия можно упростить.
Вариант пересечения задачи с самой собой у меня учтен как 
`
if (t1.equals(t2)) return false;
`
Вариант, когда время старта == null у нас, по идее, не должен появляться: в ТЗ указано, что задачи с пустой датой в приоритизированный список вообще включать не надо.
В принципе, постарался учесть твои предложения.
Спасибо!)

TyutterinYakov commented on src/dc/yandex/kanban/service/InMemoryTaskManager.java:
```diff
@@ -3,34 +3,59 @@
+
+    // Добавляет задачу в приоритизированный список
+    public void prioritize(Task task) {
+        if (task != null && task.getStartTime() != null && task.getStartTime().isAfter(Task.emptyDate)) {
```
Если время у задачи не задано - следует расположить ее в конце отсортированного списка, а их сортировку произвести по возрастанию идентификатора)

churmantey replied:
Исправил: учел это обстоятельство в компараторе.

TyutterinYakov commented on src/dc/yandex/kanban/service/FileBackedTaskManager.java:
```diff
@@ -83,7 +87,7 @@ public void deleteAllEpics() {
     // Сохраняет состояние менеджера в файл
     public void save() {
-        String header = "id,type,name,status,description,epic";
+        String header = "id,type,start_time,duration,name,status,description,epic";
```
Заголовок следует вынести в поля класса в виде константы. Так же ее можно можно сделать static, так как независимо от объекта класса заголовок в файле всегда будет один. Так же не забываем, что поля класса должны быть приватными и о том, что константы записываются заглавными буквами, а слова разделяются нижней чертой)

churmantey replied:
Исправлю. 
Насчет констант есть замечание: у меня почему-то гитхабовские автотесты не пропускают константы, написанные заглавными буквами с нижними подчеркиваниями - выдают стилевую ошибку. 
А вот если писать LCC - то все в порядке.

TyutterinYakov replied:
Возможно ты их не делаешь статическими?)

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно было бы улучшить) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

churmantey committed:
Удаление поля "тип задачи" - заменил методом.
В качестве пустой даты взят null.
Оптимизация расчета статуса и времени эпика.
Оптимизаци проверки пересечения по времени.

TyutterinYakov commented on src/dc/yandex/kanban/model/TaskStartTimeComparator.java:
```diff
@@ -0,0 +1,17 @@
+        if (o1.startTime == null && o2.startTime == null) {
+            return o1.getId() - o2.getId();
+        }
+        if (o1.startTime == null) return 1;
```
Чтобы задачи без времени старта сортировались по возрастанию идентификатора, следует немного изменить компаратор, добавив условие в начало:
`if (o1.startTime == null &&o2.startTime == null) {
     return o1.getId().compareTo(o2.getId()); }
`

churmantey replied:
так не будет работать: 
У меня id - это примитивный int. 
compareTo у него нет.

TyutterinYakov commented on src/dc/yandex/kanban/model/TaskStartTimeComparator.java:
```diff
@@ -0,0 +1,17 @@
+        }
+        if (o1.startTime == null) return 1;
+        if (o2.startTime == null) return -1;
+        if (o1.startTime.isBefore(o2.startTime)) return -1;
```
Также получать доступ к значениям полей - лучше через геттеры)
А модификаторы изменить на `private`, так как более принято делать классы максимально закрытыми)

churmantey replied:
Принято, на геттеры заменил. 
А вот приватным я поле startTime делать не стал, чтобы в сабтаске и эпике можно было использовать это поле, унаследованное от таски. 
Это плохое решение?

TyutterinYakov commented on src/dc/yandex/kanban/service/InMemoryTaskManager.java:
```diff
@@ -3,34 +3,62 @@
+        LocalDateTime t2Start = t2.getStartTime();
+        LocalDateTime t2End = t2.getEndTime();
+
+        if (t1Start == null || t2Start == null) return false;
```
Лучше вернуть просто `true`, так как мы подошли к концу списка и далее пересечение не произойдет)
Тем более, что теперь компаратор готов к `null` значениям даты)

churmantey replied:
Тут я не очень понял - это же функция проверки пересечения по времени, а не компаратор.
Здесь true возвращается, когда задачи пересекаются по времени. 
Если время хотя бы одной из них null - значит пересечения нет.

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей спринта :) Отметил еще парочку улучшений, но думаю, что ты справишься с ними вне ревью) Успехов в обучении и прекрасного дня!🌅🎇

churmantey committed:
Исправил обращение к полю времени на геттер.
